internal/component/loki/process/stages/sampling.go

```diff
@@ -22,13 +22,13 @@ var (
 
 // SamplingConfig contains the configuration for a samplingStage
 type SamplingConfig struct {
-	DropReason   *string `alloy:"drop_counter_reason,attr,optional"`
+	DropReason   string  `alloy:"drop_counter_reason,attr,optional"`
 	SamplingRate float64 `alloy:"rate,attr"`
 }
 
 func (s *SamplingConfig) SetToDefault() {
-	if s.DropReason == nil || *s.DropReason == "" {
-		s.DropReason = &defaultSamplingpReason
+	if s.DropReason == "" {
+		s.DropReason = defaultSamplingpReason
```

Comment:
I dug a little bit to check the reason behind this: what's happening is that `SetToDefault()` is always called before the args are set (that makes sense: first the defaults are set and then the params set by the user will override the defaults). The problem here was that the default was a global var and because DropReason was a pointer, its default value was the address of `defaultSamplingpReason`, not its value. So when the value is overridden by the user input, it's actually overriding the value of the var `defaultSamplingpReason`.
When the next component is built, it will also point to `defaultSamplingpReason` and modify the value with the user input. So both components are using the same value `defaultSamplingpReason` that they override.

Looking at the other stages, it seems like this is the only one that has this problem

Reply:
A while back we switched to defaultVals being a func that returned a new object for this type of reason. Glad the other stages arent affected.